src/stcal/resample/resample.py

```diff
@@ -0,0 +1,1506 @@
+            self._create_output_template_model()
+
+        else:
+            # build output WCS and calculate ouput image shape
+            if "pixel_scale" in wcs_pars and wcs_pars['pixel_scale'] is not None:
+                self._pixel_scale_ratio = wcs_pars["pixel_scale"] / input_pscale0
+                log.info(f'Output pixel scale: {wcs_pars["pixel_scale"]} arcsec.')
+                log.info(f'Computed output pixel scale ratio: {self._pixel_scale_ratio}.')
+            else:
+                self._pixel_scale_ratio = wcs_pars.get("pixel_scale_ratio", 1.0)
+                log.info(f'Output pixel scale ratio: {self._pixel_scale_ratio}')
+                self._output_pixel_scale = input_pscale0 * self._pixel_scale_ratio
+                wcs_pars = wcs_pars.copy()
+                wcs_pars["pixel_scale"] = self._output_pixel_scale
+                log.info(f'Computed output pixel scale: {self._output_pixel_scale} arcsec.')
+
+            w, ps = self._compute_output_wcs(**wcs_pars)
+            self._output_wcs = w
+            self._output_pixel_scale = ps
+            self._output_array_shape = self._output_wcs.array_shape
+            self._create_output_template_model()
+
+        # Check that the output data shape has no zero length dimensions
+        npix = np.prod(self._output_array_shape)
+        if not npix:
+            raise ValueError(
+                f"Invalid output frame shape: {tuple(self._output_array_shape)}"
+            )
+
+        assert self._pixel_scale_ratio
+        log.info(f"Driz parameter kernel: {self.kernel}")
+        log.info(f"Driz parameter pixfrac: {self.pixfrac}")
+        log.info(f"Driz parameter fillval: {self.fillval}")
+        log.info(f"Driz parameter weight_type: {self.weight_type}")
+
+        self.check_memory_requirements(allowed_memory)
+
+        log.debug('Output mosaic size: {}'.format(self._output_wcs.pixel_shape))
+
+    @property
+    def output_model(self):
+        return self._output_model
+
+    def process_kwargs(self, kwargs):
+        """ A method called by ``__init__`` to process input kwargs before
+        output WCS is created and before output model template is created.
+
+        Returns
+        -------
+        kwargs : dict
+            Unrecognized/not processed ``kwargs``.
+
+        """
+        return {k : v for k, v in kwargs.items()}
+
+    def _warn_extra_kwargs(self, kwargs):
+        for k in kwargs:
+            log.warning(f"Unrecognized argument '{k}' will be ignored.")
+
+    def check_memory_requirements(self, allowed_memory):
+        """ Called just before '_pre_run_callback()' is called to verify
+        that there is enough memory to hold the output. """
+        if allowed_memory is None and "DMODEL_ALLOWED_MEMORY" not in os.environ:
+            return
+
+        allowed_memory = float(allowed_memory)
+
+        # get the available memory
+        available_memory = psutil.virtual_memory().available + psutil.swap_memory().total
```

Comment:
if `psutil` is a dependency here, why is `bytes2human` repeated here, instead of being inherited from `psutil`?

Reply:
because `psutil` does not have this function.

Reply:
whoops, sorry